### DISCUSSION TITLE
What is your experience of connectivity and data rate issues in flight?

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1645] [Q&A]
I've put together a small barebones "ground station" app for displaying F' packets in the raw. I would find it more useful for development purposes. It's still pretty early on, but I have immediately found that the amount of telemetry being pumped into my list box locks up the programme immediately. I'm sure the locking up is something I can deal with, but maybe not the mental experience of that much data. I did a small test, and basic F' plus whatever channels my colleagues are sending (not much at this stage, I'd imagine), totals 18kB/s. Scheduled telemetry in my previous experience with flight software topped out at a few housekeeping packets every 10 seconds.

Do I take it that F' is designed to work this way, with a constant stream of channel values being sent? I know there's a packetised telemetry component lurking somewhere, but from the discussions here I gathered it to be incomplete or deprecated. Is that the case?

How does this work on Ingenuity, or other F' installations that have actually flown? Is it the case that all this data being sent by default is **intended** to be switched off? Isn't this a lot of data to be streaming to Earth when it doesn't even take into account the gazillion sensors and instruments on an actual satellite or robot?

Does your ground segment rely upon a lot of small packets updating channel values on screen at home at a high rate?

**Comment by SterlingPeet** [comment]
Let’s think about this for a minute.

**Pondering F Prime and what it Offers**

First, as a new user of F Prime, it can be very disconcerting to start up a deployment and GDS, and not see the telemetry values display in a seemingly real-time way.

Second, telemetry is a highly mission specific operations concern.

Third, the F Prime team should provide a generic solution for telemetry return, and for practical reasons it needs to be “flexible” as in covers all the normal data type cases, and it needs to be “maintainable” which means the most generic solution has the least code surface and thus requires the least engineering time to maintain.  Oh, and it has to _work_.   That’s not the same as being optimized for a specific situation.

It would seem that the provided telemetry channel system and its single packet-per-telemetry-reading architecture conveniently supports all of these things in a reasonable way.  On top of that, it delivers a working starting point to a team that is at the beginning of their flight software development journey.

Let’s look at two specific examples.

**GT-1**: [Dashboard Link](https://dashboard.satnogs.org/d/8_kvXR-nk/gt-1?orgId=1&refresh=1m)

  I wrote a lot of the drivers and OSAL for the GT-1 mission, and I intimately understand how it works, so it is an easy example for me to analyze in this context.  GT-1 did not use the built-in F Prime frame at all for telemetry.  What we did was define a telemetry struct that contained every telemetry channel we wanted, and split it over 3 packets.  Then we wrote a beacon component that packet those structs with data from the telemetry database every 30 seconds, and send those packets to the radio.

   GT-1 generates a lot more telemetry than that onboard the system, but we just don’t downlink it.  As a relatively simple mission, we didn’t do anything fancy.  There was not storage of telemetry for downlink later, or any sort of rate adjustments aside from the beacon interval setting.

   As you can see, we implemented our own version of a pattern where telemetry data is packed more efficiently than one data point per packet.  This also means that the stock F Prime GDS does not work with GT-1’s data format.  But then it also requires custom radio adapters on the flight side and the GDS side, so this is not really all that unreasonable.

**Mars Helicopter** _Because you mention this mission…_

_Caveat Emptor_: I do not work at JPL and I do not have direct insight into the helicopter’s inner workings.  But, I know some people.

Since I can’t tell you the details, I’ll provide some educated guesses about this mission:

- Lots of telemetry is generated
- None of the operations are done via “joysticking” everything is sequenced and choreographed ahead of time
- Data must travel through several “hops” so it doesn’t go directly from helicopter to the DSN, other systems relay the data first
- Communications only happens for a short period of time each day
- Any data that gets returned from helicopter had to be stored for transfer during the communications window

So, if I was designing a system like this, I would want to direct my telemetry streams to a storage device.  In a standard Linux computer this would be a hard drive, but on an embedded system it might be flash memory or (hopefully not for space applications) an SD card.

I would also want to have parameters that I can twiddle.  I want to control the rates of data that are being stored.  I want to control which channels are stored or which ones are ignored.  Lastly, I want to control the beginning and ending times of interest.

It would be reasonable to write a component that stores all this telemetry to a file (or some facsimile thereof), instead of dumping all the telemetry into the radio and hoping somebody captures it.  Then, the file can be downlinked using the provide file downlinking component in the F Prime framework, and sent to the operations team.

The same person who wrote (or at least specified) the formatting of the telemetry file should write (or at least specify) the file decoding tool that reconstitutes the telemetry data from the file into some usable format.  That could literally be plotting the telemetry channels for the operators to analyze.

 This stuff is all expected development for a mission, and way beyond the scope of inclusion for a demo of a framework’s usefulness.

But, why is this surmised solution interesting?  Well, for one, the generation of telemetry data is completely decoupled from the process of transferring it to the operators.  A big block of data is effectively cached, and then that cached data is transferred through the network as a single [large] data product to be unpacked on the ground asynchronously.

**GDS**

In general, professionally engineered GDS pipelines are optimized and customized to the mission.  F Prime does not provide a GDS that is expected to be used in a true space mission context, it is meant to work out of the box.  It is perfect for building a demo to test a component, or a few components in a development environment.  It works great for running tests of hardware without needing to spin up a fully fledged mission-ready GDS.  It’s also convenient for use when your mission’s radio isn’t really set up or ready to use.


**_In Conclusion_**

F Prime is very flexible, but it is not a complete flight system that only requires a few configurations.  You will need to develop solutions that match your needs, and F Prime is very good at supporting that.  So, I’ll ask some follow-up questions:

- What is it that you are trying to do?  Display data with your own GDS? Live?
- Have you looked at rate limiting or decimating the telemetry data?
- Do you need high telemetry data rates?
- Do you need to send every channel, or is filtering them an option?
- Do you need to see the data in real-time, or just record it and analyze it later?

The needs you have will drive the approach, and F Prime supports the option to develop solutions for wildly different needs where generic solutions will not have appropriate optimization to be performant in every scenario.

**Reply by SterlingPeet** [reply]
As a quick item to tack-on, directly addressing the data rate concern:  missions typically have data budgets that are tied to their radio link.  If the need for more data is very high, sometimes that drives the radio choice to one that supports higher data rates.

Within a robotic system, there is no direct need to generate telemetry data at the same rate that it is consumed by a GDS. There _is_ a need to generate data as fast as an algorithm needs to _use_ it.  It is not uncommon to generate data at very high rates, and then carefully select the channels and rates that actually get downlinked to the GDS as a function of the link data budget.

**Comment by timcanham** [comment]
A couple of answers:

1) The newer version of `Svc/TlmChan` fills the provided `ComBuffer` with as many channels that can fit, so that it uses the packet size more efficiently. @LeStarch Can comment if that has made it into a release yet, but it should be in the `devel` branch.
2) There is a `Svc/TlmPacketizer` component that was developed on Ingenuity that allows users to specify packets with a fixed set of telemetry channels for each packet ID. It also allows the user to specify a "level" for each packet. The `Svc/TlmPacketizer` component has commands to activate each level to control the number of packets being emitted. We have to add a code generator to the Cmake system to fully enable this, but we fully intend to release that soon. The GDS supplied with F Prime can decode these packets. Ingenuity was unique in that we had an instrument payload on the Rover we built to receive telemetry from the helicopter, and those packets were written to files on the Rover for later downlink. We used Python classes in the GDS to write standalone scripts to decode those packets back into telemetry, so the tools were really an adaptation of the existing F Prime GDS code. @SterlingPeet is correct; each project can decide how to architect their ground system, which is why we provide an XML dictionary so that other ground tools can read them to decode telemetry.

**Comment by JackNWhite** [comment]
Thanks, both of you - all good stuff!

@SterlingPeet, to answer your questions in one go, yes, I was trying to display the data in my own GDS - I'm just being picky about how I want the data displayed - but I was surprised by the amount of data. I don't have a particular requirement for anything yet, and all the options are still options. I just want to get a feel for how other people are doing it, and if there is an "F' Way of Telemetry".

**Reply by SterlingPeet** [reply]
That’s cool!  I would say that the “F` Way of Telemetry” is largely “have it your way” 😄 

Personally, I have found that updating the user display faster than 5 times per second requires much fancier and more difficult programming, and it usually does not provide actionable engineering data to display it that fast.  You can certainly _collect_ it faster, that usually doesn’t cause problems.  Unless you are doing critical tele-operation joysticking [or making a game 🤣 ] I would leave the faster refresh rates alone for reduced development effort.

Good luck with your project!
